tools/Environments/DevHome.Environments/Helpers/DataExtractor.cs

```diff
@@ -32,9 +33,46 @@ public static List<OperationsViewModel> FillDotButtonOperations(ComputeSystem co
             operations.Add(new OperationsViewModel("Delete", "\uE74D", computeSystem.DeleteAsync));
         }
 
-        // ToDo: Correct the function used
-        // operations.Add(new OperationsViewModel("Pin To Taskbar", "\uE718", computeSystem.));
-        // operations.Add(new OperationsViewModel("Add to Start Menu", "\uE8A9", computeSystem.));
+        return operations;
+    }
+
+    public static async Task<List<OperationsViewModel>> FillDotButtonPinOperationsAsync(ComputeSystem computeSystem)
+    {
+        var supportedOperations = computeSystem.SupportedOperations;
+        var operations = new List<OperationsViewModel>();
+        if (supportedOperations.HasFlag(ComputeSystemOperations.PinToTaskbar))
+        {
+            var pinResultTaskbar = await computeSystem.GetIsPinnedToTaskbarAsync();
+            if (pinResultTaskbar.Result.Status == ProviderOperationStatus.Success)
+            {
+                if (pinResultTaskbar.IsPinned)
+                {
+                    operations.Add(new OperationsViewModel("Unpin From Taskbar", "\uE77A", computeSystem.UnpinFromTaskbarAsync));
+                }
+                else
+                {
+                    operations.Add(new OperationsViewModel("Pin To Taskbar", "\uE718", computeSystem.PinToTaskbarAsync));
+                }
+            }
+        }
+
+        if (supportedOperations.HasFlag(ComputeSystemOperations.PinToStartMenu))
+        {
+            var pinResultStartMenu = await computeSystem.GetIsPinnedToStartMenuAsync();
+            if (pinResultStartMenu.Result.Status == ProviderOperationStatus.Success)
+            {
+                if (pinResultStartMenu.IsPinned)
+                {
+                    operations.Add(new OperationsViewModel("Unpin From Start Menu", "\uE77A", computeSystem.UnpinFromStartMenuAsync));

```

Comment:
Won't this crash Dev Home if it this ComputeSystem doesn't implement IComputeSystem2? Even if a developer messes up and needs to implement that, we should not let them crash Dev Home.

Reply:
We use a wrapper class called `ComputeSystem` for all IComputeSystem objects retrieved from an extension. He added a method in that class here: https://github.com/microsoft/devhome/pull/2636/files#diff-b42364b144fb64835ea4dc7177d09887cf83790c3a55bfc029e1b7b6195f534aR298

so we should be good here. Calling it will just return a ComputeSystemOperationResult that with a failure ProviderOperationResult